src/test/java/com/uber/cadence/workflow/WorkflowTest.java

```diff
@@ -1848,67 +1863,63 @@ public void mySignal(String value) {
-                  Optional.of(execution.get().getRunId())));
-          assertEquals("Hello ", client.get().getState());
-
-          // Test getTrace through replay by a local worker.
-          Worker queryWorker;
-          if (useExternalService) {
-            Worker.Factory workerFactory = new Worker.Factory(service, DOMAIN);
-            queryWorker = workerFactory.newWorker(taskList);
-          } else {
-            queryWorker = testEnvironment.newWorker(taskList);
-          }
-          queryWorker.registerWorkflowImplementationTypes(TestSignalWorkflowImpl.class);
-          String queryResult = null;
-          try {
-            queryResult =
-                queryWorker.queryWorkflowExecution(
-                    execution.get(), "QueryableWorkflow::getState", String.class);
-          } catch (Exception e) {
-            throw new RuntimeException(e);
-          }
-          assertEquals("Hello ", queryResult);
-          try {
-            Thread.sleep(500);
-          } catch (InterruptedException e) {
-            throw new RuntimeException(e);
-          }
-          client.get().mySignal("World!");
-          assertEquals("World!", client.get().getState());
-          assertEquals(
-              "Hello World!",
-              workflowClient
-                  .newUntypedWorkflowStub(execution.get(), Optional.empty())
-                  .getResult(String.class));
-        });
-    client.set(workflowClient.newWorkflowStub(QueryableWorkflow.class, optionsBuilder.build()));
+    QueryableWorkflow client =
+        workflowClient.newWorkflowStub(QueryableWorkflow.class, optionsBuilder.build());
     // To execute workflow client.execute() would do. But we want to start workflow and immediately
     // return.
-    execution.set(WorkflowClient.start(client.get()::execute));
+    WorkflowExecution execution = WorkflowClient.start(client::execute);
+
+    sleep(Duration.ofSeconds(1));
+    assertEquals(workflowId, execution.getWorkflowId());
+    // Calls query multiple times to check at the end of the method that if it doesn't leak threads
+    int queryCount = 100;
+    for (int i = 0; i < queryCount; i++) {
+      assertEquals("initial", client.getState());
+    }
+    sleep(Duration.ofSeconds(1));
+
+    client.mySignal("Hello ");
+    sleep(Duration.ofSeconds(1));
+
+    // Test client created using WorkflowExecution
+    QueryableWorkflow client2 =
+        workflowClient.newWorkflowStub(
+            QueryableWorkflow.class, execution.getWorkflowId(), Optional.of(execution.getRunId()));
+    assertEquals("Hello ", client2.getState());
+
+    String queryResult = null;
+    try {
+      queryResult =
+          queryWorker.queryWorkflowExecution(
+              execution, "QueryableWorkflow::getState", String.class);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    assertEquals("Hello ", queryResult);
+    sleep(Duration.ofMillis(500));
+    client2.mySignal("World!");
```

Comment:
Why are we using 2 workflow clients?

Reply:
Just to ensure that a client created from workflow id also works correctly.